homeassistant/components/anglian_water/coordinator.py

```diff
@@ -0,0 +1,58 @@
+"""DataUpdateCoordinator for anglian_water."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import Any
+
+from pyanglianwater import AnglianWater
+from pyanglianwater.exceptions import (
+    ExpiredAccessTokenError,
+    InvalidPasswordError,
+    InvalidUsernameError,
+    ServiceUnavailableError,
+    UnknownEndpointError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryError
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, LOGGER
+
+
+class AnglianWaterDataUpdateCoordinator(DataUpdateCoordinator[dict[str, Any]]):
+    """Class to manage fetching data from the API."""
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: AnglianWater,
+    ) -> None:
+        """Initialize."""
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(minutes=20),
+        )
+        self.client = client
+
+    async def _async_update_data(self):
+        """Update data via library."""
+        try:
+            await self.client.update()
+        except InvalidUsernameError as exception:
+            raise ConfigEntryError(exception) from exception
+        except InvalidPasswordError as exception:
+            raise ConfigEntryError(exception) from exception
+        except UnknownEndpointError as exception:
+            raise UpdateFailed(exception) from exception
+        except ServiceUnavailableError as exception:
+            raise UpdateFailed(exception) from exception
+        except ExpiredAccessTokenError:
```

Comment:
What kind of token is it? Do we know what the expiration is?

Reply:
its a bearer token (not jwt though), usually expires after 20 mins

Reply:
Sounds like they maybe just have set times when they invalidate a bunch?

Reply:
If we only fetch yesterday, is 20 minutes not a lot of polling?

Reply:
> Sounds like they maybe just have set times when they invalidate a bunch?

maybe, I'm not too sure, this API is reverse engineered from their mobile app (which requires you to login every time you open it). AW are aware of this integration though as I have spoken to one of their directors about a separate security vulnerability - off the back of this, depending on politics the API might change but for now smart water meters aren't regulated in the UK and therefore no common API exists.

> If we only fetch yesterday, is 20 minutes not a lot of polling?

20 mins is actually a side effect of copying the code from my custom integration, that uses the meter readings to feed the statistics like opower does. Sometimes readings can be available for the current day too, but all this goes back to my earlier comment about there being no standards in the UK for this so it varies. I will add this in a future PR once we've got this in :)